File: GO-Leet/124. Binary Tree Maximum Path Sum.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxPathSum(root *TreeNode) int {
	ans := math.MinInt32

	var helper func(*TreeNode) int
	helper = func(root *TreeNode) int {
		if root == nil {
			return 0
		}

		left := max(helper(root.Left), 0)
		right := max(helper(root.Right), 0)
		ans = max(ans, left+right+root.Val)

		return max(left, right) + root.Val
	}

	helper(root)

	return ans
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: 124-binary-tree-maximum-path-sum/124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxPathSum(TreeNode root) {
       maxPathSumDownFrom (root);
    return a ;
        
  }

  private int a = Integer.MIN_VALUE;

  // root->val + 0/1 of its subtrees
    
  private int maxPathSumDownFrom(TreeNode root)
  {
    if (root == null)
      return 0;

    final int l = Math.max(maxPathSumDownFrom(root.left), 0);
    final int r = Math.max(maxPathSumDownFrom(root.right), 0);
      
    a = Math.max(a, root.val + l + r);
      
    return root.val + Math.max(l, r);
  } 
    }
